### PR TITLE
MAINT: Import time: Avoid importing python.Threading

### PR DESCRIPTION
**Pull request by hmaarrfk**
See here:
https://github.com/numpy/numpy/pull/14083

Just testing on CIs for now.

### CODE REVIEW

**Comment by hmaarrfk** [comment]
@eric-wieser do you know why azure won't run on this? I'm pretty sure CPython will work, just not PyPy.

**Comment by seberg** [comment]
I think some heisenbug or so on windows came up a bit often on azure the last few PRs? (not quite sure though).

**Comment by hmaarrfk** [comment]
Yeah, I think so too.

File: numpy/random/bit_generator.pyx [hunk 1]
```diff
@@ -43,13 +43,12 @@ except ImportError:
...
-try:
-    from threading import Lock
-except ImportError:
-    from dummy_threading import Lock
-
 from cpython.pycapsule cimport PyCapsule_New
 
+# Threading isn't the cheapest import. Import from the lightweight
+# _thread module,
+from _thread import allocate_lock as Lock
```
Comment (seberg) [inline numpy/random/bit_generator.pyx]: It seems to me that the `_thread` method might be choking the pypy test? (i.e. that lock waits forever?). How OK is it really to use the `_thread` name?

Comment (hmaarrfk) [reply numpy/random/bit_generator.pyx]: The python STL was documenting it.

But I guess I misunderstood what they meant by low level.

https://docs.python.org/3/library/concurrency.html

I kinda knew I was doing a bad thing, so I didn't try it first. I think I might have to rerun initial benchmarks now that some of the other PRs are going through

Comment (seberg) [reply numpy/random/bit_generator.pyx]: Dunno, it seems that python might have started to endorse `_thread` specfically, and possibly pypy just has a bug.

Comment (hmaarrfk) [reply numpy/random/bit_generator.pyx]: I have to start rerunning these benchmarks again. Some of these `fixes` are related to deprecation cleanups. Others like this one are more hazy. One could argue that importing from `_thread` is obscure and shouldn't be done even if it was supported by both PyPy.

**Comment by eric-wieser** [comment]
How does this behave on systems whether threading is unavailable?

**Comment by hmaarrfk** [comment]
Oh I guess that is what dummy threading was for. Not too sure for now. PyPy has threading right?

**Comment by seberg** [comment]
But you can compile python without threading. PyPy should fail the import if it had no threading, but it seems rather that it hangs on an infinite Lock somewhere.

File: numpy/core/_internal.py [hunk 2]
```diff
@@ -17,7 +16,7 @@
 except ImportError:
     ctypes = None
 
-IS_PYPY = platform.python_implementation() == 'PyPy'
+IS_PYPY = 'PyPy' in sys.version
```
Comment (mattip) [inline numpy/core/_internal.py]: Please use `'__pypy__' in sys.modules` which is more robust

**Comment by mattip** [comment]
I am not sure about this PR. It seems to increase code fragility for a minimal decrease in import time.

File: numpy/core/_internal.py (hunk 2 above)
Comment (hmaarrfk) [reply numpy/core/_internal.py]: maybe, but the code snippet above was taken from CPython
https://github.com/python/cpython/blob/master/Lib/platform.py#L1032
So L

Comment (mattip) [reply numpy/core/_internal.py]: meh. If I  file an issue / make a PR there to change that to something more robust, how will we know to resync the code here?

Comment (mattip) [reply numpy/core/_internal.py]: ok, I am convinced  :)

File: numpy/random/bit_generator.pyx (hunk 1 above)
Comment (hmaarrfk) [reply numpy/random/bit_generator.pyx]:
```
root@c766aae4e273:/# pypy3.6-v7.1.1-linux64/bin/pypy3 
Python 3.6.1 (784b254d6699, Apr 14 2019, 10:22:42)
[PyPy 7.1.1-beta0 with GCC 6.2.0 20160901] on linux
Type "help", "copyright", "credits" or "license" for more information.
And now for something completely different: ``snow, snow''
>>>> import _thread
>>>> from _thread import allocate_lock as Lock
>>>> Lock
<built-in function allocate_lock>
```
This seems to work on the latest downloaded PyPy

**Comment by hmaarrfk** [comment]
I think it might have a more measurable effect once https://github.com/numpy/numpy/pull/14097 gets pulled in. I was able to get down to 70 ms in my original omnibus PR. I can bechmark after that gets approved again.

**Comment by hmaarrfk** [comment]
It might be difficult to see the results of this without getting rid of a `shutil` import in `_datasource`. `shutil` ends up importing many different compression formats trying to see if they are installed in a pythonic way at import time.

Unfortunately, `bz2` imports `Threading` just for `RLock`. I think a PR can be made upstream to use the lightweight `_thread` module there too.
https://github.com/python/cpython/blob/master/Lib/bz2.py#L17

**Comment by hmaarrfk** [comment]
I'm not sure upstream will accept using `_thread.RLock` as they specifically created a workaround for it.

https://github.com/python/cpython/blob/master/Lib/threading.py#L95

That said, even pypy has RLock implemented in C, so maybe :/.

It might be more fruitful to fix upstream. They don't seem to have a c implementation of `excepthook` which is causing `traceback` to be imported when the `threading` module is imported.
https://github.com/python/cpython/blob/master/Lib/threading.py#L1128

**Comment by hmaarrfk** [comment]
FYI, it seems that the above comment is under development, and might actually be fixed in python 3.8.

**Pull request closed without merging**